In [20]:
%%html

<script>
  function code_toggle() {
    if (code_shown){
      $('div.input').hide('500');
      $('#toggleButton').val('Show Code')
    } else {
      $('div.input').show('500');
      $('#toggleButton').val('Hide Code')
    }
    code_shown = !code_shown
  }

  $( document ).ready(function(){
    code_shown=false;
    $('div.input').hide()
  });
</script>
<form action="javascript:code_toggle()"><input type="submit" id="toggleButton" value="Show Code"></form>

# Introduction to Python

This notebook demonstrates the introductory python programming skills by writing sequential algorithms to solve problems involving input, processing and output (IPO) approach.

## 1.1: Set Up Your Python Environment
* Download Python: [Link](https://www.python.org/downloads/) 
* Install Python: [Instruction](http://introtopython.org/programming_environment.html)

## 1.2: Basic : Python Programming
Programming is the way of teaching a computer to interect with a user. We are going to start this by saying "Welcome to Python World!".

In [2]:
print("Welcome to Python World!")

Welcome to Python World!


### 1.2.1: Variables and Input Reading
**Variables**: Variables are simply some placeholders in the computer memory for some value. The values can be anything like numbers (integers, floats, etc.), strings (text), chars (single letters) etc.

We can store values in variables and use them later. Suppose, you have 1 pen and 2 pencils. Here are some variable assignments:

pen = 1

pencil = 2

Variable values can (and will) change over time, let’s say your friend give you another pen. Now the value of pen variable will be:

pen = pen + 1

pen = 1 + 1

pen = 2

**Reading Input from the Keyboard**: Almost in all programs, we need to read some values from the user as the input, and store them in some variables, so that we can use them later.

variable = input("string prompt message")

color = input("Your favourite color:" )

The input function always returns a string, even if the user enters numeric data. And, we can not use strings for mathematical operations. So we need to convert strings to numeric data. Here is how we do this. The int(string input) function, converts the string input to numeric data. The same applies for the float() function.

age = int(input("Your age :")

In [24]:
color = input("your favourite color: ")

age = int(input("your age : "))

## 1.3: Algorithm
An algorithm is a set of operations or steps to be performed to solve a problem or a task. Let's write a simple algorithm of our daily morning life:

1. Wake up from sleep
2. Brush your teeth
3. Take a shower
4. Take the breakfast
5. Get into the car or bus for school

Let's write another simple algorithm to call to your friend:

1. Take your mobile phone
2. Go to your contact list
3. Search your friend's name
4. Call your friend

These two algorthims are the simple examples that we follow in our real life. In computing, programmers write the set of instructions of an algorithm that instructs computer to perform that task.

### 1.3.1: Sequential Algorithm

An sequential algorithm performs the set of operations sequentially from start to finish. This algorithm can be solved using input/processing/output (IPO) approach. In sequential algorithm, processing occurs only when all input required data is available and output occurs only after appropriate processing has occurred. Let's see the following example that will give the clear idea of input, processing and output of a sequential algorithm:

**Example 1**

***Write an algorithm or program that will take numbers from you/user, and show the average of these three number.***

**Solution:**

To solve this algorithm or program, we need three numbers which are the inputs of the algorithm and the equation $(num1 + num2 + num3) / 3$ will be act as processing unit. The algorithm will not be proceed to processing stage unless all the inputs are available and without processing unit we will not get the output unit. So. this algorithm can be solved using IPO approach and all processes will be performed from start to finish. So, this is also a sequential algorithm.

**Algorithm:**

In [3]:
import io
import base64
from IPython.display import HTML

video = io.open('Video/avg_flowchart.mp4', 'r+b').read()
encoded = base64.b64encode(video)
HTML(data='''<video alt="test" controls>
                <source src="data:video/mp4;base64,{0}" type="video/mp4" />
             </video>'''.format(encoded.decode('ascii')))

In [8]:
#
# Compute the average
#

# Read the input value from user : 3 numbers
num1 = int(input("Enter your number 1: "))
num2 = int(input("Enter your number 2: "))
num3 = int(input("Enter your number 3: "))

# Compute the avaerage of the numbers : Processing
avg = (num1 + num2 + num3)/3

# Display the result : Output
print("The average of the numbers:", avg)

The average of the numbers: 5.0


**Testing**

Let's test the above program using the following test cases:

Input = (0, 5, 7) -----> Output = 4

Input = (0, -2, 5)-----> Output = 1

Input = (1.5,  , ) -----> Error will be shown after taking the first number as input is taken as interger format, it will not accept any decimal point float number.

**Revised Program**

In [25]:
#
# Compute the average
#

# Read the input value from user : 3 number
num1 = float(input("Enter your number 1: "))  #Change occurs
num2 = float(input("Enter your number 2: "))  #Change occurs
num3 = float(input("Enter your number 3: "))  #Change occurs

# Compute the avaerage of the numbers : Processing
avg = (num1 + num2 + num3)/3

# Display the result : Output
print("The average of the numbers:", avg)

The average of the numbers: 5.1000000000000005


The above example has been solved using the IPO approach. Let's assume another problem as read the three numbers and display these numbers at the output. But this problem can not be solved using the IPO approach because there is no processing unit in this problem. Input unit will take three numbers and output unit will just display these numbers. Let's see another example of sequential algorithm that can be solved using IPO approach.

**Example 2**

***Write an algorithm or program that will read seconds and then displays the equivalent hours, minutes remaining seconds.***

**Solution:**

We can solve this problem by taking seconds as input, performing suitable logics or equations to get equivalent hours, minutes, remaining seconds as processing unit and finally display the result as output unit. This is also an example of a sequential algorithm as all the stages are performed from starting to finishing point.

**Algorithm**

In [2]:
import io
import base64
from IPython.display import HTML

video = io.open('Video/time_flowchart.mp4', 'r+b').read()
encoded = base64.b64encode(video)
HTML(data='''<video alt="test" controls>
                <source src="data:video/mp4;base64,{0}" type="video/mp4" />
             </video>'''.format(encoded.decode('ascii')))

In [11]:
#
# Time conversion : Seconds to equivalent hours, minutes and remaining seconds
#

# Read the seconds from user : Input
second = int(input("Enter the time in seconds: "))

# Compute the hours: Processing
hour = second/3600
second = second - (hour * 3600)

# Compute the minutes and remaining seconds: Processing
minute = second/60
second = second - (minute * 60)

# Display the result of Hours, Minutes and Seconds : Output
print("Equivalent Time:", hour , "hours", minute, "minutes", second, "seconds")

Equivalent Time: 1.025 hours 7.579122514774402e-15 minutes 0.0 seconds


**Testing**

Let's test the above program using the following test cases:

1. Case 1: Input = (3600) -----> Output = 1 hours 0 minutes 0 seconds

2. case 2: Input = (3620) -----> Output = 1.005555 hours 0 minutes 0 seconds, But the output should be 1 hours 0 minutes 20 seconds

3. Case 3: Input = (3690) -----> Output = 1.025 hours 7.57912e-15 minutes 0 seconds, But the output should be 1 hours 1 minutes 30 seconds

Case 1 and 2 are not giving the expected outcomes. As the goal of program is to convert the seconds into hours, minutes and seconds, so hour and minute can not be decimal point and both of them will be integer. 

** Revised Program**

In [12]:
#
# Time conversion : Seconds to equivalent hours, minutes and remaining seconds
#

# Read the seconds from user : Input
second = int(input("Enter the time in seconds: "))

# Compute the hours: Processing
hour = int(second/3600)         # Change occurs
second = second - (hour * 3600)

# Compute the minutes and remaining seconds: Processing
minute = int(second/60)        # Change occur
second = second - (minute * 60)

# Display the result of Hours, Minutes and Seconds : Output
print("Equivalent Time:", hour , "hours", minute, "minutes", second, "seconds")

Equivalent Time: 1 hours 1 minutes 30 seconds


### Exercise 1
Write a program that will calculate total cost of a restaurant bill. The program will read initial cost of the bill. Tax rate is $10\%$, and tip is $5\%$ of the initial cost (without tax). The program displays the output as tax, tip and overall cost of the meal. 

Test the program by taking the following inputs:
1. initial cost: 120
2. initial cost: 150.5

In [13]:
#
# Restaurant bill : Compute the overall cost, tax and tip amount
#

TAX_RATE = .1
TIP_RATE = .05

# Read the seconds from user : Input
initial_cost = float(input("Enter the initial cost: "))

# Compute the tax, tip and total: Processing
tax = initial_cost * TAX_RATE         
tip = initial_cost * TIP_RATE
overall_cost = initial_cost + tax + tip

# Display the result of tax, tip and overall cost : Output
print("Tax amount:", tax , "\nTip amount:", tip, "\nOverall Cost:", overall_cost)

Tax amount: 56.800000000000004 
Tip amount: 28.400000000000002 
Overall Cost: 653.1999999999999


### Exercise 2
Write a program that will take a two digit number and display the sum of its two digits. 

Test the program using the following inputs:
1. number: 10
2. number: 15

In [14]:
#
# Adding the digits of a two digit number
#

# Read the number from user : Input
num = int(input("Enter the number: "))

# Compute the tens, ones and add them: Processing
tens = int(num/10)         
ones = num - (tens * 10)
add = tens + ones

# Display the result of Hours, Minutes and Seconds : Output
print("Tens digit:", tens , "\nOnes digit:", ones, "\nSum", add)

Tens digit: 1 
Ones digit: 5 
Sum 6
